octeon: add support for Ubiquiti EdgeRouter 4

damex opened the pull request:
Ubiquiti EdgeRouter 4 is 4 port Octeon Cavium 7130 powered router.
It has internal power supply and needs c13 power cord.
There are three 10/100/1000 Mbps RJ45/Copper ports and
one 1000 Mbps SFP port connected directly to a SoC.

SoC:
  Octeon Cavium 7130 (Cavium 3)
  Clocked at 1000Mhz

Memory:
  1 GiB (SK hynix H5TQ4G63CFR-RDC × 2)
  DDR3, clocked at 533 Mhz (1066Mhz effective)
Flash:
  - mtd:
    8 MiB (Macronix MX25L6408EMI-12G)
    used for uboot/eeprom
  - emmc:
    4 GiB (SanDisk SDIN7DP2-4G)
    used for kernel+rootfs

Leds: 1x for power status (white/blue, controllable)
  and 4x for ethernet and sfp ports (no control over them)
Buttons: 1x Reset (from SOC)
Serial: 1x RJ45 port on front panel. 115200 baud, 8N1 (from SoC)
USB: 1x USB3.0 on front panel (from SoC)
MII: 1x QSGMII from SoC is used
PHY: 1x Vitesse VSC8504 of which 4x ports is used

All physical port numbers are properly mapped inside OS and
named by lanX instead of ethX.

There is also special purpose four(4) loopX ports available.
That loopX ports are currently hardcoded by linux kernel
and exact use case of them is currently unknown. We leave them
to the linux kernel and octeon board defaults.

All four (4) physical ports are connected to the same QSGMII.
vsc8504 is used for phys and only 4, 5, 6 and 7 phys are used.

Phy mapping:
 - Phy5 is connected to physical eth0 port
 - Phy6 is connected to physical eth1 port
 - Phy7 is connected to physical eth2 port
 - Phy4 is connected to physical eth3 port

Why this device needs external dts:
 - faster boot time since need to initialize less device tree nodes.
 - to add actual indication with LED about boot/failure/upgrade.
   i.e. user could know when to enter failsafe mode or if upgrade is done
 - reset button support so user can reset their device in case off failure
 - sfp port indication in dmesg with information about sfp module
   it also indicates when module inserted or removed

Octeon quirks:
 - There is no port status available before it interface brought up
 - SFP port can not be tied to actual phy due to octeon-ethernet state
   and currently we can only get reports a about SFP state in dmesg

How to flash the firmware:
  - copy openwrt-octeon-ubnt_edgerouter-4-initramfs-kernel.bin and
    openwrt-octeon-ubnt_edgerouter-4-squashfs-sysupgrade.tar to
    USB flash drive that is formatted to vfat/fat32
  - connect USB flash drive to edgerouter 4 front USB port
  - connect serial cable using front RJ45 port (115200 baud, 8N1)
  - connect power to cable to edgerouter 4
  - connect terminal to the console to see uboot boot process
  - interrupt boot by pressing button(s) on your keyboard to log in to the uboot
  - detect usb connected flash drives by typing to the console:
    usb start
  - after drive is detected load initramfs+kernel to the memory by typing:
    fatload usb 0:1 0x20000000 openwrt-octeon-ubnt_edgerouter-4-initramfs-kernel.bin
  - after initramfs+kernel is loaded to the memory load it by typing:
    bootoctlinux 0 numcores=4 endbootargs mem=0
  - boot process should finish and you will be greeted with console after pressing enter
  - create directory to mount usb flash drive to by typing:
    mkdir /tmp/sda
  - mount flash drive to that directory by typing:
    mount /dev/sda1 /tmp/sda
  - flash firmware to router internal storage by typing:
    sysupgrade /tmp/sda/openwrt-octeon-ubnt_edgerouter-4-squashfs-sysupgrade.tar
  - device will reboot and after it gets up you will have edgerouter 4 running openwrt

Reviewed-by: Johannes Kimmel <fff@bareminimum.eu>
Tested-by: Johannes Kimmel <fff@bareminimum.eu>
Signed-off-by: Roman Kuzmitskii <damex.pp@icloud.com>

Commit:
octeon: add support for Ubiquiti EdgeRouter 4

Ubiquiti EdgeRouter 4 is 4 port Octeon Cavium 7130 powered router.
It has internal power supply and needs c13 power cord.
There are three 10/100/1000 Mbps RJ45/Copper ports and
one 1000 Mbps SFP port connected directly to a SoC.

SoC:
  Octeon Cavium 7130 (Cavium 3)
  Clocked at 1000Mhz

Memory:
  1 GiB (SK hynix H5TQ4G63CFR-RDC × 2)
  DDR3, clocked at 533 Mhz (1066Mhz effective)
Flash:
  - mtd:
    8 MiB (Macronix MX25L6408EMI-12G)
    used for uboot/eeprom
  - emmc:
    4 GiB (SanDisk SDIN7DP2-4G)
    used for kernel+rootfs
Leds: 1x for power status (white/blue, controllable)
  and 4x for ethernet and sfp ports (no control over them)
Buttons: 1x Reset (from SOC)
Serial: 1x RJ45 port on front panel. 115200 baud, 8N1 (from SoC)
USB: 1x USB3.0 on front panel (from SoC)
MII: 1x QSGMII from SoC is used
PHY: 1x Vitesse VSC8504 of which 4x ports is used

All physical port numbers are properly mapped inside OS and
named by lanX instead of ethX.

There is also special purpose four(4) loopX ports available.
That loopX ports are currently hardcoded by linux kernel
and exact use case of them is currently unknown. We leave them
to the linux kernel and octeon board defaults.

All four (4) physical ports are connected to the same QSGMII.
vsc8504 is used for phys and only 4, 5, 6 and 7 phys are used.

Phy mapping:
 - Phy5 is connected to physical eth0 port
 - Phy6 is connected to physical eth1 port
 - Phy7 is connected to physical eth2 port
 - Phy4 is connected to physical eth3 port

Why this device needs external dts:
 - faster boot time since need to initialize less device tree nodes.
 - to add actual indication with LED about boot/failure/upgrade.
   i.e. user could know when to enter failsafe mode or if upgrade is done
 - reset button support so user can reset their device in case off failure
 - sfp port indication in dmesg with information about sfp module
   it also indicates when module inserted or removed

Octeon quirks:
 - There is no port status available before it interface brought up
 - SFP port can not be tied to actual phy due to octeon-ethernet state
   and currently we can only get reports a about SFP state in dmesg

How to flash the firmware:
  - copy openwrt-octeon-ubnt_edgerouter-4-initramfs-kernel.bin and
    openwrt-octeon-ubnt_edgerouter-4-squashfs-sysupgrade.tar to
    USB flash drive that is formatted to vfat/fat32
  - connect USB flash drive to edgerouter 4 front USB port
  - connect serial cable using front RJ45 port (115200 baud, 8N1)
  - connect power to cable to edgerouter 4
  - connect terminal to the console to see uboot boot process
  - interrupt boot by pressing button(s) on your keyboard to log in to the uboot
  - detect usb connected flash drives by typing to the console:
    usb start
  - after drive is detected load initramfs+kernel to the memory by typing:
    fatload usb 0:1 0x20000000 openwrt-octeon-ubnt_edgerouter-4-initramfs-kernel.bin
  - after initramfs+kernel is loaded to the memory load it by typing:
    bootoctlinux 0 numcores=4 endbootargs mem=0
  - boot process should finish and you will be greeted with console after pressing enter
  - create directory to mount usb flash drive to by typing:
    mkdir /tmp/sda
  - mount flash drive to that directory by typing:
    mount /dev/sda1 /tmp/sda
  - flash firmware to router internal storage by typing:
    sysupgrade /tmp/sda/openwrt-octeon-ubnt_edgerouter-4-squashfs-sysupgrade.tar
  - device will reboot and after it gets up you will have edgerouter 4 running openwrt

Reviewed-by: Johannes Kimmel <fff@bareminimum.eu>
Tested-by: Johannes Kimmel <fff@bareminimum.eu>
Signed-off-by: Roman Kuzmitskii <damex.pp@icloud.com>

damex commented:
https://gist.github.com/damex/4819288cf8132434b318616ec38151c0 openwrt flash log.

damex commented:
i just pushed update about renaming interfaces from ethX to a properly matched (to physical) order using lanX naming.

please help with reviewing MR.

adschm commented:
Please add flashing instructions to your commit message.

adschm commented on target/linux/octeon/base-files/lib/preinit/01_sysinfo:
```diff
@@ -17,6 +17,11 @@ do_sysinfo_octeon() {
 		;;
 
+	"UBNT_E300"*)
+		# let 02_sysinfo handle it since it has its own device tree
```
Please make it more clear that you are referring to _generic_ 02_sysinfo here.

damex replied:
sure, adjusted commentary.

adschm commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -224,3 +224,15 @@ macaddr_canonicalize() {
+		[ -d "/sys/class/net/$device/device" ] && \
+		[ -f "/sys/class/net/$device/address" ] && \
+		[ "$(cat /sys/class/net/$device/address)" = "$interface_mac" ] && {
+			ip link set "$device" name "$interface_name"
```
This will break if multiple devices have the same MAC address?

damex replied:
- it ignores 'virtual devices' that have same mac as physic (bridges, vlans and such).
- it works only for <actual device> that have device directory that describe the actual device and that have mac address.
- it expects <actual devices> to have unique mac addresses or function does not work for them

so yes, if let's say person have 4 ethX devices that have the same mac address (i.e. `DE:AD:BE:EF:00:00`) - that function won't work for them

another example - if person has 4 ethX devies that have unique mac addresses (i.e. `DE:AD:BE:EF:00:00`, `DE:AD:BE:EF:00:01`, `DE:AD:BE:EF:00:02` and `DE:AD:BE:EF:00:04`) - that function will work for them.

adschm replied:
Okay, that's what I assumed. So, since MAC addresses being not unique happens frequently, a function relying on uniqueness should not be added. After all, I really don't think this mechanism will bring us forward.
I'd very much prefer to just hardcode the interface rename without the detour of using MAC addresses.

damex replied:
>  So, since MAC addresses being not unique happens frequently, a function relying on uniqueness should not be added.

mac addresses are unique across the octeon target.

> I'd very much prefer to just hardcode the interface rename without the detour of using MAC addresses.

kernel does not guarantee interface naming.
what if order change because of any other reason? that would break logic here.

renaming using mac address imho would be a much more resilient solution for the problem here compared to renaming using interface name since interface names might change in the future but mac addresses board initialize with is known and will not be changed.

adschm replied:
> mac addresses are unique across the octeon target.

How can you say that? Do you know the MAC address assignment of all devices people use with generic image? Every vendor does it differently ...
The solution could work for your special limited case, but I still don't like to establish such a mechanism which will be used by others then and create problems elsewhere.

> kernel does not guarantee interface naming order.
> what if order change because of any other reason? that would break logic here.

Then we would update and provide a migration script.

Unless order is known to vary with the same exact kernel (i.e. be random intrinsically), I'd prefer to bind to interface names.

You might also split this discussion from the device support itself by using interface names for now, and then add the MAC address-based assignment later in separate patch.

damex replied:
>then add the MAC address-based assignment later in separate patch.

damex replied:
> You might also split this discussion from the device support itself by using interface names for now, and then add the MAC address-based assignment later in separate patch.
if we

damex replied:
> How can you say that? Do you know the MAC address assignment of all devices people use with generic image? Every vendor does it differently ...

every single device octeons tree have (octeon+) - have device tree filled with static board specific mac addresses. that is their ‘feature’.
 there is only few boards in a tree and they are fine to use rename by mac since their mac addresses are mapped in device tree already.

if there is no generic board currently that don’t have mac set.

and yeah, we are not forcing people to use that rename feature.
if board fit the requirements - that rename feature could be used. if it is not - it would not. it is as simple as that.

> Then we would update and provide a migration script.

it sounds like ‘create a problem and provide a solution’ to me instead of providing a resilient solution.

i disagree about using device names to rename because that solution will be very fragile.

we better not rename devices at'all at this point of time if renaming by mac creates so much trouble.

i will update a MR removing rename feature completely and we will stick with whatever kernel provides.

damex replied:
so what should we do to interface rename?
i am okay with removing that feature altogether (instead of adding a band aid to rename by interface name) if that is gonna make merge happen.

adschm commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -224,3 +224,15 @@ macaddr_canonicalize() {
 	printf "%02x:%02x:%02x:%02x:%02x:%02x" 0x${canon// / 0x} 2>/dev/null
 }
+
+rename_interface_by_mac() {
```
Since only octeon is using that for now, I would keep it in the corresponding 03_rename_interfaces until it's required elsewhere.

damex replied:
sure, thanks. moved function to `03_rename_interfaces`.

adschm commented on target/linux/octeon/base-files/lib/preinit/03_rename_interfaces:
```diff
@@ -0,0 +1,15 @@
+rename_interfaces() {
+	. /lib/functions/system.sh
+	. /lib/functions.sh
```
From my memory it's more common to have the includes outside of the function. I'd prefer that.

damex replied:
sure, moved include(s) outside of the function.

adschm commented on target/linux/octeon/base-files/lib/preinit/03_rename_interfaces:
```diff
@@ -0,0 +1,15 @@
+
+	case "$(board_name)" in
+		ubnt,edgerouter-4)
+			rename_interface_by_mac lan0 $(macaddr_add $(mtd_get_mac_binary eeprom 0) +1)
```
Please use 0x0 instead of 0 for consistency with similar statements elsewhere (location/offset should always be hex).

damex replied:
sure, used hex instead.

adschm commented on target/linux/octeon/base-files/lib/preinit/03_rename_interfaces:
```diff
@@ -0,0 +1,15 @@
+			rename_interface_by_mac lan0 $(macaddr_add $(mtd_get_mac_binary eeprom 0) +1)
+			rename_interface_by_mac lan1 $(macaddr_add $(mtd_get_mac_binary eeprom 0) +2)
+			rename_interface_by_mac lan2 $(macaddr_add $(mtd_get_mac_binary eeprom 0) +3)
+			rename_interface_by_mac lan3 $(mtd_get_mac_binary eeprom 0)
```
What's the actual reason for doing a "rename" of interfaces here? What's wrong with the default?

lemmi replied:
Without this, the order does does not correspond to the labels on the device, which is very confusing. Any other attempts to fix this in the DTS failed so far.

adschm replied:
But why that complicated detour with the MAC addresses? Can't we just rename eth3 to lan2, eth9 to lan1 etc.?
So, instead of determining the names by the MAC addresses _every_ time, just determine it once and put the result here?

damex replied:
because we need a resilient solution and currently kernel does not guarantee interface naming order.

maybe we could do that if we opt for https://en.wikipedia.org/wiki/Consistent_Network_Device_Naming but openwrt does not use that.

renaming by mac address is much more resilient solution.
from my experience it is even more resilient then that 'consistent network device naming' from kernel. i have had devices change names between kernel releases or reboots (because of device initialization order) but mac addresses obviously still stay the same.

adschm replied:
Have you tried already to just patch in a label like done recently for mt7621/DSA:
https://github.com/openwrt/openwrt/commit/3211c983fd7a51208f22866193df64a009b64fdb

damex replied:
thanks, just added a patch to do it via device tree labels

adschm commented on target/linux/octeon/base-files/lib/upgrade/platform.sh:
```diff
@@ -87,6 +92,9 @@ platform_do_upgrade() {
 		;;
+	ubnt,edgerouter-4)
+		kernel=mmcblk0p1
+		;;
```
This can be merged with er?

damex replied:
yes, seems like it could be.

thanks, merged it with `er`.

adschm commented on target/linux/octeon/config-5.4:
```diff
@@ -25,13 +25,8 @@ CONFIG_BLK_MQ_PCI=y
-CONFIG_CAVIUM_OCTEON_LOCK_L2_MEMCPY=y
-CONFIG_CAVIUM_OCTEON_LOCK_L2_TLB=y
+CONFIG_CAVIUM_OCTEON_CVMSEG_SIZE=0
+# CONFIG_CAVIUM_OCTEON_LOCK_L2 is not set
```
This changes quite a lot of options. I'd like to have that in a separate commit.

Probably this should be tested with other devices as well.

damex replied:
sure, i dropped all the changes regarding disabling l2 cache locking.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "cn7130.dtsi"
```
I don't see a point in creating this additional cn7130.dtsi. The compatible will be overwritten anyway, so it's actually a no-op.

lemmi replied:
There is work on the way to support more than just the ER-4, namely the ER-12. That one still needs work because of the integrated switch chip. This will also make it more convenient to get the ER-6p running, once we have access to that board.

damex replied:
i dropped `cn7130.dtsi` for now.
`cn7130_ubnt_edgerouter-4.dts` now include universal `cn71xx.dtsi`
it is not needed now. we can always add it late if needed with extra customization (when they will actually be needed).

adschm replied:
> it is not needed now. we can always add it later if extra customization is needed (when it actually is needed).

Yes, thanks. If there a reason to have them, I'm fine with having multiple nested DTSI files later.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+		led-failsafe = &led_power_blue;
+		/* Constant Blue */
+		led-running = &led_power_blue;
+		/* Blue + Blinking White */
```
Can you explain how these patterns are constructed?

damex replied:
default settings is gpio15 (blue) high + gpio17 (white) high results in constant white + blue pattern.

> /* Constant Blue */
> led-running = &led_power_blue;

settings gpio15 (blue) low + gpio17 (white) high results in constant blue - that is how we get constant blue.

> /* White + Blinking Blue */
> led-boot = &led_power_white;

if we start triggering trigger white led we - get blinking blue light that looks like white led is there but blue is blinking.

> /* Blue + Blinking White */
> led-failsafe = &led_power_blue;
> /* Blue + Blinking White */
> led-upgrade = &led_power_blue;

if we start triggering trigger blue led we - get blinking white light that looks like blue led is there but white is blinking.

also there is this bonus (not used here):
if we set gpio15 (blue) low + gpio17 (white) low - it results in power light being completely off.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+		led_power_blue: power_blue {
+			label = "blue:power";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
```
Please always add an empty line between nodes.

damex replied:
sure, added spaces between phy and led nodes.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+
+	keys {
+		compatible = "gpio-keys-polled";
+		/* Poll interval time in milliseconds */
```
drop comment

damex replied:
dropped.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		boot0: partition@0 {
```
You can drop the DT labels (boot0:) for this one and dummy, as they are not used anywhere.

damex replied:
sure, dropped the labels.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+		eeprom: partition@400000 {
+			label = "eeprom";
+			read-only;
+			reg = <0x400000 0x10000>;
```
Sure this is only 0x10000?

damex replied:
yes, it is only 0x10000.
used stock firmware for the reference and it seem to contain 'all the eeprom' in there.

adschm commented on target/linux/octeon/dts/cn7130_ubnt_edgerouter-4.dts:
```diff
@@ -0,0 +1,213 @@
+		compatible = "vitesse,vsc8504", "ethernet-phy-ieee802.3-c22";
+		reg = <4>;
+		sfp = <&sfp>;
+	};
```
empty lines between nodes.

damex replied:
thanks, added.

adschm commented on target/linux/octeon/image/Makefile:
```diff
@@ -47,6 +51,19 @@ define Device/ubnt_edgerouter
+  DEVICE_MODEL := EdgeRouter 4
+  SOC := cn7130
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_DTS := $$(SOC)_$(1)
```
If defined explicitly anyway, I'd prefer just the string here, as there is no benefit in using the variables anymore.

damex replied:
sure, dropped `SOC` variable since it is not used anywhere outside `DEVICE_DTS` and defined `DEVICE_DTS` statically. also reused `DEVICE_DTS` in `KERNEL_DEPENDS` and `append-dtb-to-elf` function

lemmi commented:
Tested:

- [x] building
- [x] booting Initramfs
- [x] flashing from Initramfs
- [x] sysupgrade -c
- [x] LEDs
- [x] Interface renaming
- [x] SFP port
- [x] SFP module detection
- [x] reset button for reboot
- [x] reset button for factory reset
    - LEDs start flashing blue/white after a long press
    - correctly resets

damex commented:
> Please add flashing instructions to your commit message.

added flashing instructions to commit message and did a rebase.
please check it again.

adschm commented on target/linux/octeon/base-files/lib/preinit/03_rename_interfaces:
```diff
@@ -0,0 +1,26 @@
+	done
+}
+
+rename_interfaces() {
```
Consider changing to "octeon_rename_interfaces" ...

damex replied:
renamed to `octeon_rename_interfaces`

adschm commented on target/linux/octeon/patches-5.4/150-octeon_e300_support.patch:
```diff
@@ -0,0 +1,18 @@
+ 		ENUM_BRD_TYPE_CASE(CVMX_BOARD_TYPE_CUST_PRIVATE_MIN)
+ 		ENUM_BRD_TYPE_CASE(CVMX_BOARD_TYPE_UBNT_E100)
+ 		ENUM_BRD_TYPE_CASE(CVMX_BOARD_TYPE_ITUS_SHIELD)
++		ENUM_BRD_TYPE_CASE(CVMX_BOARD_TYPE_UBNT_E300)
```
Any reason for that sorting? (Same above ...)

damex replied:
kernel use sorting by board id.
that is the only reason.

adschm replied:
Ah okay, fine with me.

adschm commented on target/linux/octeon/image/Makefile:
```diff
@@ -47,6 +51,18 @@ define Device/ubnt_edgerouter
+define Device/ubnt_edgerouter-4
+  DEVICE_VENDOR := Ubiquiti
+  DEVICE_MODEL := EdgeRouter 4
+  DEVICE_DTS_DIR := ../dts
```
Does octeon already have DTS files upstream or is this mach-files only?
In the first case, we should use the same directory as kernel does.

damex replied:
technically octeon have dts files available upstream but for much (much!) older targets.
https://github.com/torvalds/linux/tree/master/arch/mips/boot/dts/cavium-octeon
there is nothing related to cavium3 devices (related cn71xx.dtsi here).
 
is there any other way to make that happen beside making device tree files as a patches for linux kernel?
that would be huge step back in ability to maintain target devices.

adschm replied:
The idea would be to just use directory "files/arch/mips/boot/dts/cavium-octeon" then instead of "dts".

Everything in "files" or "files-5.4" will be just added to the local kernel tree. This is used on many targets, e.g. https://github.com/openwrt/openwrt/tree/master/target/linux/ipq40xx/files/arch/arm/boot/dts

damex replied:
do i need to do something for `files` to be added to local kernel tree? basically i just tried moving files from `dts` dir to `files/arch/mips/boot/dts/cavium-octeon`, and no files from `files` directory is copied to local kernel tree.

adschm replied:
Normally, this just works out-of-the-box. I'm not sure to which directory of the various directories containing the kernel it is copied exactly then.

Note that you might need adjustments in image/Makefile for building, though. This would done with target-specific DTS_DIR then, not with device-specific DEVICE_DTS_DIR...

damex replied:
@adschm seems like for it to work i need to add a `generic` subtarget. that is the only difference between octeon and other targets and adding it - make `files` being able to be copied. if i remove it - files are not copied. will it be okay to add `generic` subtarget ?

damex replied:
@adschm i found that files from <target>/files directory is copied to kernel directory only once when you do a fresh kernel build. rebuilds does not carry new files over. this is a huge way back from separate `dts` directory.

damex replied:
i juar migrated device tree files to `files/arch/mips/boot/dts/cavium-octeon` and rebased against it.

adschm replied:
> i found that files from target/files directory is copied to kernel directory only once when you do a fresh kernel build. rebuilds does not carry new files over. 

That's strange, because for others targets it typically just worked to pull a PR adding a new DTS file, select it in make menuconfig and build.

damex commented:
rebased to utilize new octeon_move_config for config moving in preinit

adschm commented on target/linux/octeon/image/Makefile:
```diff
@@ -7,13 +7,18 @@
 
 define Device/Default
   PROFILES = Default $$(DEVICE_NAME)
+  DTS_DIR := $(DTS_DIR)/cavium-octeon
```
DTS_DIR is a target variable, not a device variable. So, it should be assigned outside of Device/Default.

damex replied:
would it be suitable to add DTS_DIR to <target>/Makefile ?

damex replied:
or maybe we can use `DEVICE_DTS_DIR := $(DTS_DIR)/cavium-octeon` in `Device/Default` instead?
what is preferred?

Commit:
octeon: implement interface rename by label from device tree

adds patch to octeon ethernet driver that to rename interface
name by label from device tree

Tested-by: Johannes Kimmel <fff@bareminimum.eu>
Signed-off-by: Roman Kuzmitskii <damex.pp@icloud.com>

damex commented:
need more help reviewing changes.

@adschm helped reviewing changes

@lemmi helped testing changes

adschm commented:
@Grommish This patch touches several kernel config options. Can you possibly have a look and/or check whether that affects Itus Shield?

damex commented:
note about options - we get to enable `CONFIG_STMMAC_ETH` only to get to `CONFIG_SFP` since there is no other simpler and/or easier option to get to `CONFIG_SFP`. this haven't changed even with last 5.9 or 5.10 kernels.

no device in octeon tree use `CONFIG_STMMAC_ETH` directly and there is no problem with devices having extra kernel options.

damex commented:
i just enabled `CONFIG_NET_DSA` instead of `CONFIG_STMMAC_ETH` (we only used it in this MR to get to `CONFIG_SFP`) to get to `CONFIG_SFP` (it is needed for SFP cage support) option and to reuse that option later for other octeon devices with switches that we plan to add.

also enabled `CONFIG_SFP` as a kernel module (package) instead of building it into the kernel since not all devices have SFP cage.

Commit:
octeon: honor disabled sgmii device tree nodes

adds patch to octeon ethernet driver that lets sgmii interface
device tree node to be disabled and that disabled interface
won't be unnecessarily initialized.

It solves the problem with octeon boards that have 8 sgmii or more ports
initialized but have nothing connected to them.

Tested-by: Johannes Kimmel <fff@bareminimum.eu>
Signed-off-by: Roman Kuzmitskii <damex.pp@icloud.com>

damex commented:
i just added a patch `161-honor-sgmii-node-device-tree-status.patch` that helps board to not initialize unneded device tree nodes for SGMII interfaces.

we need it for octeon_ethernet to stop initializing all (possibly available) 8 SGMII interfaces on er4 when there is only 4 connected physically and enabled/configured via device tree.

https://gist.github.com/damex/cfefee53172528c5f8c81dd31f4490c4 correct output.

lemmi commented:
Tested and reviewed again with the new changes. Unused Interfaces are gone. LGTM

Grommish commented:
None of these changes effect the Itus device.  I built/flashed from https://github.com/damex/openwrt/tree/add_ubiquiti_edgerouter_4_support with defconfig - no issues.

I'm good with it :)

Commit:
generic: add support for vsc8504 phy

adds support for vsc8504 phy.
patch have use on 5.4 kernel and has
to be dropped after since phy is supported by
CONFIG_MICROSEMI_PHY on newer LTS kernels.

Tested-by: Johannes Kimmel <fff@bareminimum.eu>
Signed-off-by: Roman Kuzmitskii <damex.pp@icloud.com>

damex commented:
thanks to everyone efforts it is finally got merged.

we retested it again after merge and it works well on edgerouter lite, itus shield and newly introduced edgerouter 4.

adschm commented:
@damex @lemmi Can this "octeon: honor disabled sgmii device tree nodes" patch be upstreamed?

damex commented:
@adschm
i think it should be possible for both patches.

`octeon: honor disabled sgmii device tree nodes`
`octeon: implement interface rename by label from device tree `

planning on attempting to upstream this patches but haven't had a chance to do it (yet, quite busy).

adschm commented:
I also wonder whether we can do something with that probe order patch, originally introduced in 2014:
https://github.com/openwrt/openwrt/commit/f083ca97435183b08711b6d4afa4e80faeb78b0f

Unfortunately, there is no comment available there.

adschm commented:
140-octeon_e300_support.patch should also be upstreamable if that one is:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/mips/include/asm/octeon/cvmx-bootinfo.h?id=90761cee5f4c9a97e9952e1a8e2165c8ab346ea4

damex commented:
> I also wonder whether we can do something with that probe order patch, originally introduced in 2014:
> [f083ca9](https://github.com/openwrt/openwrt/commit/f083ca97435183b08711b6d4afa4e80faeb78b0f)
> 
> Unfortunately, there is no comment available there.

it is edgerouter pro / edgerouter 8. now it is in broken state (disabled by default) and proper fix would be to drop that patch and use device tree fixing its behavior.

adschm commented:
I will send 140-octeon-e300 patch myself. It's simple enough ...
